[1단계 - 출석] 우가(민보욱) 미션 제출합니다.

bowook opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

--> https://prolog.techcourse.co.kr/studylogs/3955

## 체크 리스트

- [ ] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

<!-- 이유 작성 -->
충족을 완벽하게 하지 못한 가장 큰 요인은 엔티티를 작게 유지하지 못한 것 같습니다.
그리고 3개 이상의 인스턴스 변수를 가지는 클래스를 가지고 있기 때문에 객체지향 생활체조 요구사항을 올바르게 충족했다고 할 수 없을 것 같아요.


  
## 어떤 부분에 집중하여 리뷰해야 할까요?

<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 피드백받고 싶은 내용을 공유해주세요.  
예를 들어, 가장 고민했던 점이나 여전히 어려운 부분, 그리고 이에 대한 생각을 적을 수 있습니다. -->

1. 라인수를 10라인으로 하라는데, 이 라인 시작점을 어디를 기준으로 해야하는지 고민입니다.
-> 시그니쳐를 시작으로 보는 것인지, 그 내부를 시작으로 보는 것인지 궁금합니다.

2. 메서드 분리를 정말 간단한 기능인데도 분리를 해야하는지 궁금합니다.

3. Date, Time 등을 묶은 DateInfo 너무 많은 도메인을 묶는 과정에서 Depth가 깊어지는 것이 아닌가? 라는 생각이 들었어요.
-> 그러면 outputView에서 .get .get .get 이러한 상황이 생기는데, view에서는 get을 사용해도 된다라고 저번 단계 피드백에서 받았는데, 저렇게 연속된 .get도 허용을 하는 것인지 궁금합니다.

bowook committed:
docs(README.md) : 리드미 초안 작성

bowook committed:
feat(Crew) : 크루 객체 구현

bowook committed:
test(CrewTest) : 크루 객체 테스트

bowook committed:
build : 테스트 라이브러리 추가

bowook committed:
docs(attendances.scv) : 초기 파일 셋팅

bowook committed:
docs(README.md) : 구현한 기능 체크

bowook committed:
feat(Application) : 의존성 주입

bowook committed:
feat(AttendanceMachine) : 컨트롤러 구현

bowook committed:
feat(AttendanceStatus) : 출석 상태 기능 추가

bowook committed:
feat(Crew) : Crew 일급 컬렉션 사용 추가

bowook committed:
test(CrewsTest) : CrewsTest 테스트 코드 작성

bowook committed:
feat(CrewStatus) : CrewStatus 크루 상태 기능 추가

bowook committed:
test(CrewStatusTest) : CrewStatusTest 크루 상태 테스트 코드 작성

bowook committed:
feat(Exception) : 예외 처리 기능 추가

bowook committed:
feat(View) : 입력, 출력 기능 추가

bowook committed:
feat(DateInfo) : DateInfo 출석 세부 정보 기능 추가

bowook committed:
test(DateInfoTest) : DateInfoTest 출석 세부 정보 테스트 코드 작성

bowook committed:
feat(DateInfos) : DateInfos 출석부 역할 기능 추가

bowook committed:
test(DateInfosTest) : DateInfosTest 출석부 역할 테스트 코드 작성

bowook committed:
feat(DayOfWeek) : 날짜 값 한국어로 변환하는 역할 기능 추가

bowook committed:
test(DayOfWeekTest) : 날짜 값 한국어로 변환하는 테스트 코드 작성

bowook committed:
feat(FileReader) : 파일 읽어오기 기능 추가

bowook committed:
feat(Function) : 선택할 수 있는 기능 목록 추가

bowook committed:
test(FunctionTest) : 선택할 수 있는 기능 목록 테스트 코드 작성

bowook committed:
feat(Time) : 시간 정보 저장하는 기능 추가

bowook committed:
test(TimeTest) : 시간 정보 테스트 코드 작성

bowook committed:
feat(Register) : 출석 시스템 기능 추가

ddaaac commented on src/main/java/attendance/controller/AttendanceMachine.java:
```diff
@@ -0,0 +1,132 @@
+        Crews crews = Crews.fromCrewsFile(FileReader.fileReadCrewNames("attendances.csv"));
+        LocalDate now = LocalDate.now();
+        Register register = new Register(crews, now);
+        register.fromCrewAttendanceTimeFile(crews, FileReader.fileReadLine("attendances.csv"));
```
I/O는 CPU 연산에 비해 꽤 무거운 작업입니다. 같은 파일이라면 한 번만 읽는게 좋을 것 같아요.

bowook replied:
한 번만 사용하도록 수정하겠습니다!

ddaaac commented on src/main/java/attendance/controller/AttendanceMachine.java:
```diff
@@ -0,0 +1,132 @@
+
+    public void start() throws IOException {
+        Crews crews = Crews.fromCrewsFile(FileReader.fileReadCrewNames("attendances.csv"));
+        LocalDate now = LocalDate.now();
```
- 요구사항에는 2024년 12월에만 사용하는 것처럼 나와있는데 이렇게하면 2025년 2월까지 출결이 모두 표시되는 걸까요?
- `now()` 관련 함수들을 사용할 때는 꼭 타임존을 명시해주세요. 내가 만든 프로그램이 어떤 환경에서 실행될지는 아무도 모릅니다. 만약 UTC나 미국 시간을 기준으로 하는 환경에서 실행된다면 의도하지 않은 동작을 할 수도 있어요.

bowook replied:
- 의도한 방식은 현재 날짜 까지의 출석 상태를 보여주어야 한다고 생각을 하여서 now를 사용한 것입니다. 그래야 출석부 역할을 이행하는 것이 아닌가? 라는 생각을 하였어요.
- 아,, 타임존을 고려하지 못했네요. 당연히 한국 시간 기준이라고 생각을 하였습니다. 수정하겠습니다!

ddaaac commented on src/main/java/attendance/domain/constant/Function.java:
```diff
@@ -0,0 +1,27 @@
+import attendance.exception.ErrorMessage;
+import java.util.Arrays;
+
+public enum Function {
```
java에는 Function 클래스가 존재해서 헷갈리 수 있는 네이밍인 것 같아요. 또, 좀 더 구체적인 네이밍을 하면 알기도 편하겠죠.

bowook replied:
네이밍 수정하겠습니다!.. 당시 너무 급하다 보니까 클래스명, 메서드명을 수정하지 못했네요!

ddaaac commented on src/main/java/attendance/domain/Register.java:
```diff
@@ -0,0 +1,71 @@
+            String day = make(date, "-", 2);
+            Time time = Time.from(timeNumber);
+            modifyInfo(crew, Integer.parseInt(day), time);
+        }
```
여기 메서드를 보고 어떤 로직이 돌아가는지 파악이 쉽지 않네요. 메서드 이름이나 상수에 이름을 부여한다던가 다양한 방식으로 개선해보시면 좋을 것 같아요.

bowook replied:
상수에 이름을 부여해서 의미를 직관적으로 표시해줄 수 있도록 수정하겠습니다!

ddaaac commented on src/main/java/attendance/controller/AttendanceMachine.java:
```diff
@@ -0,0 +1,132 @@
+
+    private boolean mappingFunction(Function function, LocalDate now, Crews crews, Register register) {
+        if (function.equals(Function.ONE)) {
+            functionOne(now, crews, register);
```
functionOne같은 이름보다는 구체적으로 어떤 일을 하는지 나타낼 수 있는 네이밍을 사용해볼까요?

bowook replied:
출석 확인 기능의 역할을 수행하는 것이니까, checkAttendance와 같은 직관적인 네이밍으로 수정해볼게요!

ddaaac commented on src/main/java/attendance/controller/AttendanceMachine.java:
```diff
@@ -0,0 +1,132 @@
+        register.fromCrewAttendanceTimeFile(crews, FileReader.fileReadLine("attendances.csv"));
+
+        boolean flag = true;
+        while (flag) {
```
마찬가지로 flag 보다는 어떤 진행상태인지를 나타내는 변수명을 사용하면 좋을 것 같아요.

bowook replied:
Quit이 입력받는 순간 종료가 되니까, isQuit?, isRunning? 으로 수정해보겠습니다!

ddaaac commented on src/main/java/attendance/domain/Time.java:
```diff
@@ -0,0 +1,64 @@
+
+public class Time {
+    private final String hour;
+    private final String minute;
```
LocalTime이란 표준 객체 대신 직접 Time을 만들어주신 이유가 있을까요?

bowook replied:
Time이 초기 출석 상태인 "--:--"을 가지고 있어야 하고, 사용자가 출석 시간이나 수정할 시간을 입력할 때도 숫자가 아닌 다른 문자나 형식에 맞지 않는 값을 입력했을 때 예외를 발생시키는 역할이 있다고 생각을 해서 TIme을 만들었습니다.

ddaaac commented on src/main/java/attendance/domain/constant/DayOfWeek.java:
```diff
@@ -0,0 +1,34 @@
+    SUNDAY(7, "일요일"),
+    ;
+
+    private final int day;
```
여기에 java DayOfWeek를 갖고 있어도 될 것 같네요

bowook replied:
아 맞네요! 어차피 dayOfWeek의 .getValue()를 통해서 나온 값을 해당하는 요일명으로 변환해주기 위한 작업이였거든용! 감사합니다.

ddaaac commented on src/main/java/attendance/domain/constant/DayOfWeek.java:
```diff
@@ -0,0 +1,34 @@
+
+import java.util.Arrays;
+
+public enum DayOfWeek {
```
마찬가지로 자바 표준 객체와 이름을 동일하게 사용하면 나중에 헷갈립니다. 다른 네이밍을 사용해주세요.

bowook replied:
넵! 네이밍 주의하겠습니다

ddaaac commented on src/main/java/attendance/domain/DateInfo.java:
```diff
@@ -0,0 +1,124 @@
+    private final String month;
+    private final String day;
+    private final DayOfWeek dayOfWeek;
+    private Time time;
```
여기도 LocalDateTime과 같은 표준 시간 객체를 사용해보면 좋을 것 같습니다.

bowook replied:
표준 시간 객체에 대해서 알아보고 적용해보도록 하겠습니다!

ddaaac commented on src/main/java/attendance/domain/Register.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    public DateInfo modifyInfo(Crew crew, int date, Time modifyTime) {
+        DateInfos dateInfos = register.get(crew);
```
해당 크루가 없다면 어떻게 되나요?

bowook replied:
초기 attendances.csv 파일에서 크루의 이름을 다 가져와서 crews에 저장을 해 놓았습니다.
이후, 사용자로부터 crew명을 입력 받았을 때, 저장된 crews에서 findCrew를 통해 일치하는 크루를 찾아 Crew로 반환을 해줍니다.
여기서 만일 존재하지 않는 crew명을 입력할 경우 crew의 findCrew가 예외를 발생합니다.
그래서 modiyfInfo까지 전달된 crew는 존재하는 크루라고 생각을 합니다.

ddaaac replied:
Register 객체가 아닌 다른 곳에서 검증이 이뤄지고, Register 객체의 메서드를 호출하게 되니 특별히 검증할 필요가 없다라는 얘기일까요? 답변해주신 내용을 종합해보면 절차지향적인 관점으로 구현을 하신 것 같습니다.

여기도 위 코멘트와 마찬가지로 이 객체만 놓고 보면 좋을 것 같아요.
객체 지향에서는 이 객체는 언제든 다른 로직에서 사용할 수 있습니다. 따라서 이 객체만 달랑 놓고 보더라도 온전한 로직을 가져야하죠.

다른 로직에서 사용할 때 여기 전달된 crew는 존재하지 않는 크루가 될 수 있습니다. 당장 테스트 코드를 짜면서 존재하지 않는 크루 이름을 넣을 수 있겠죠.

bowook replied:
맞네요..
객체만 놓고 바라보면 존재하지 않는 크루를 넣을 수 있네요..
객체만 놓고 바라보는게 아직 잘 적응이 안되서,, 덕분에 이해할 수 있었습니다. 감사합니다.

Register에서 crew가 존재하는지 확인하는 검증 코드를 추가하였습니다!

ddaaac commented on src/main/java/attendance/domain/DateInfos.java:
```diff
@@ -0,0 +1,92 @@
+    private final List<DateInfo> dateInfos;
+    private int absence;
+    private int late;
+    private int attendance;
```
- 클래스명에 출석에 관련된 정보가 포함되면 좋겠네요
- Info와 같은 단어들은 모호합니다. "정보"라는 단어가 붙어도 그 의미를 유추하기는 쉽지 않아요.

bowook replied:
DataInfos가 출석 정보만이 아닌 좀 포괄적인 느낌의 네이밍인 것 같네요!
AttendanceRegistry로 변경하도록 하겠습니다!

ddaaac commented on src/main/java/attendance/domain/Register.java:
```diff
@@ -0,0 +1,71 @@
+        crews.register(register, DateInfos.fromDefaultValue(now));
+    }
+
+    public DateInfo modifyInfo(Crew crew, int date, Time modifyTime) {
```
이런 곳도 LocalDateTime을 받으면 파라미터 수를 줄일 수 있겠네요.

bowook replied:
감사합니다!

ddaaac commented on src/main/java/attendance/domain/DateInfo.java:
```diff
@@ -0,0 +1,124 @@
+        if (this.attendanceStatus.equals(AttendanceStatus.ABSENCE)) {
+            return 1;
+        }
+        return 0;
```
- 1 또는 0을 반환할거라면 booelan을 반환하는 건 어떨까요?
- check란 네이밍은 보통 유효성 검사를 할 때 사용됩니다. 적절하게 이름을 바꿔보는 건 어떨까요?

bowook replied:
1과 0을 반환 받아서 해당 메서드를 호출하는 부분에서 값에 즉각적으로 덧셈이 이루어지도록 하려고 의도한 것입니다.
boolean으로 반환을 하게되면 해당 반환 값이 참인지 거짓인지를 한 번더 비교를 한 후 덧셈이 이루어져야해서 int로 반환하는 것이 적절하다고 생각을 했습니다!

메서드 명은 isAbsence()로 변환하도록 하겠습니다!

ddaaac replied:
> 1과 0을 반환 받아서 해당 메서드를 호출하는 부분에서 값에 즉각적으로 덧셈이 이루어지도록 하려고 의도한 것입니다.

객제 지향적으로 생각할 때는 이 객체를 호출하는 부분이 아닌 이 객체 자체만 놓고 보는 연습을 하면 좋습니다.
`boolean isAbsence()`와 `int isAbsence()` 두 시그니처를 봤을 때 어느쪽이 DateInfo 입장에서 자연스러운 시그니처일까요?

boolean->int 변환은 이 메서드를 호출하는 클라이언트에서 처리해야 자연스럽지 않을까싶네요.

bowook replied:
감사합니다!
객체지향적인 관점이 아직 많이 미숙한 것 같다고 느껴지네요!
말씀해주신 내용을 읽어보고 고민을 많이 하였는데,
출결을 확인해주는 객체는 결국 말 그대로 출석을 했다, 출석을 안 했다에 대한 참, 거짓 값만 반환해주면 되는 책임이 있다고 생각을 하였어요.
그리고 이를 호출하는 곳에서 해당 반환된 값을 받아 로직을 수행하는 것이 적절하다고 생각이 드네요!
감사합니다!!

ddaaac commented on src/main/java/attendance/domain/DateInfo.java:
```diff
@@ -0,0 +1,124 @@
+    private AttendanceStatus checkAttendanceStatus(String hourLimit) {
+        int hourMinute = Integer.parseInt(time.getHour() + time.getMinute());
+        int absentTime = Integer.parseInt(hourLimit + "30");
+        int lateTime = Integer.parseInt(hourLimit + "05");
```
지각, 결석 등을 다루는 enum이 있으면 조금 더 30분, 5분에 대한 도메인 규칙을 빠르게 파악할 수 있을 것 같네요.

bowook replied:
enum에 30과 05에 해당하는 상수를 추가해서 해보겠습니다!

ddaaac commented on src/main/java/attendance/domain/DateInfo.java:
```diff
@@ -0,0 +1,124 @@
+    }
+
+    private boolean checkHoliday() {
+        return dayOfWeek.getDayOfWeek().equals("토요일") || dayOfWeek.getDayOfWeek().equals("일요일");
```
내부 String으로 비교하기 보다는 enum자체로 비교하면 좋겠네요.

bowook replied:
수정하겠습니다!

ddaaac commented on src/main/java/attendance/domain/DateInfo.java:
```diff
@@ -0,0 +1,124 @@
+    }
+
+    private boolean checkDefault() {
+        return time.getHour().equals("--");
```
도메인 로직이 View와 연관된 것으로 보여요.

bowook replied:
Time을 지우고 전부 LocalDateTime으로 관리할 수 있도록 수정했어요!
시간 부분은 초기에 0을 넣어주고 view에서 0이 발견되면 "-"로 출력할 수 있도록 수정했습니다.

ddaaac commented on src/main/java/attendance/domain/Register.java:
```diff
@@ -0,0 +1,71 @@
+        Map<Crew, List<Integer>> riskCrews = new HashMap<>();
+        for (Crew crew : register.keySet()) {
+            DateInfos dateInfos = register.get(crew);
+            findRiskCrews(crew, riskCrews, dateInfos);
```
정렬에 관한 요구사항이 있는데, 혹시 다른 곳에서 정렬하는 곳이 있을까요?

bowook replied:
정렬하는 부분은 미처 구현하지 못했습니다!
구현을 하게 된다면 view에서 정렬을 수행하도록 할 것 같습니다.
출력을 할 때 정렬이 된 상태로 보여줘야 하는 것이라 생각해서,
 Register에서 정렬된 상태를 가지고 있을 책임은 없다고 생각을 했습니다!

ddaaac commented on src/main/java/attendance/controller/AttendanceMachine.java:
```diff
@@ -0,0 +1,132 @@
+        inputView.closeScanner();
+    }
+
+    private boolean mappingFunction(Function function, LocalDate now, Crews crews, Register register) {
+        if (function.equals(Function.ONE)) {
+            functionOne(now, crews, register);
+            return true;
+        }
+        if (function.equals(Function.TWO)) {
+            functionTwo(crews, register);
+            return true;
+        }
+        if (function.equals(Function.THREE)) {
+            functionThree(crews, register);
+            return true;
+        }
+        if (function.equals(Function.FOUR)) {
+            functionFour(register);
+            return true;
+        }
+        return false;
```
> 라인수를 10라인으로 하라는데, 이 라인 시작점을 어디를 기준으로 해야하는지 고민입니다.
-> 시그니쳐를 시작으로 보는 것인지, 그 내부를 시작으로 보는 것인지 궁금합니다.

요건 코치분들께 여쭤보는 게 더 확실할 것 같습니다.

bowook replied:
다음 주 데일리 미팅때 여쭤봐야겠네요! 감사합니다!

ddaaac commented on src/main/java/attendance/domain/DateInfo.java:
```diff
@@ -0,0 +1,124 @@
+        }
+        return parsedNumber;
+    }
+    //TODO : 리뷰 : 라인 수 궁금, 메서드 분리 궁금
```
의문이 들었던 구체적인 곳에 코멘트 달아주시면 얘기해보기 좋을 것 같네요.

ddaaac commented on src/main/java/attendance/domain/DateInfo.java:
```diff
@@ -0,0 +1,124 @@
+        return dayOfWeek.getDayOfWeek();
+    }
+
+    public Time getTime() {
+        return time;
+    }
```
> Date, Time 등을 묶은 DateInfo 너무 많은 도메인을 묶는 과정에서 Depth가 깊어지는 것이 아닌가? 라는 생각이 들었어요.
-> 그러면 outputView에서 .get .get .get 이러한 상황이 생기는데, view에서는 get을 사용해도 된다라고 저번 단계 피드백에서 받았는데, 저렇게 연속된 .get도 허용을 하는 것인지 궁금합니다.

View를 위함이면 어쩔 수 없겠죠. 어떤 값을 보여주거나, 저장할 때는 객체지향을 지킬 수 없습니다. 객체지향을 지킬 필요도 없구요.

bowook replied:
감사합니다! 엄청 궁금한 부분이었어요!

ddaaac reviewed (changes requested):
안녕하세요 우가 👋 

몇 가지 코멘트 남겼으니 확인해주세요.

bowook committed:
refactor(AttendanceMachine) : 파일 I/O 한 번만 하도록 수정

bowook committed:
refactor(AttendanceMachine) : now를 사용함으로써 타임존 설정

bowook committed:
refactor(CommandOption) : Function 네이밍 수정 및 메서드 명 수정

bowook committed:
refactor(Register) : 상수화를 통한 의미 전달할 수 있도록 수정

bowook committed:
refactor(Weekday) : 클래스 명 수정, int day값이 아닌 DayOfWeek 을 가지고 있도록 수정

bowook committed:
refactor(DateInfo) : 메서드 명 수정

bowook committed:
refactor(DateInfo) : 문자열로 비교하는 것이 아닌 enum 자체로 비교하도록 수정

bowook committed:
refactor(AttendanceRegistry) : DateInfos -> AttendanceRegistry로 클래스 명 수정

출석 관련 정보를 표시해주기 위함

bowook committed:
refactor : LocalDateTime을 사용하도록 수정하였습니다.

bowook committed:
refactor : AttendanceStatus에 상수를 추가했습니다.

bowook committed:
refactor(OutputView) : 뷰에서 Hyphen과 0에 대해 처리하도록 수정

bowook committed:
refactor : 정렬

ddaaac commented on src/main/java/attendance/domain/Crews.java:
```diff
@@ -0,0 +1,46 @@
+    public static Crews fromCrewsFile(List<String> lines) {
+        Set<Crew> crewNames = new HashSet<>();
+        for (String line : lines) {
+            crewNames.add(Crew.from(List.of(line.split(",")).get(0)));
```
`getFirst()`를 활용할 수 있겠네요.

bowook replied:
인덱스를 통해 접근하는 것 보다 더 직관적인 방법인 것 같아요! 적용하겠습니다!

ddaaac commented on src/main/java/attendance/controller/AttendanceMachine.java:
```diff
@@ -0,0 +1,136 @@
+        Crews crews = Crews.fromCrewsFile(lines);
+        LocalDate now = LocalDate.now(ZoneId.of("Asia/Seoul"));
+        Register register = new Register(crews, now);
+        register.fromCrewAttendanceTimeFile(crews, lines);
```
lines에서 이름 추출부분과 시간 추출 부분을 분리할 필요가 있나요 싶어요.
지금은 Crews, Register 객체에서 String -> 객체 변환 책임을 가져가고 있는데 이게 적당한지도 고민해볼만 할 것 같네요.

List<String>을 Register로 만들어주는 유틸성 클래스가 있어도 좋겠다는 생각이 드네요.

bowook replied:
lines라는 같은 데이터를 다른 방식으로 두 번 가공하는 것이 저도 고민을 하고 있었습니다. 
말씀해주신 방식대로 유틸성 클래스를 만들어서 처리하면 데이터를 객체로 변환하는 역할을 한 곳에서 처리할 수 있을 것 같다는 생각이 드네요!,
한 번 적용해보겠습니다! 감사합니다!

ddaaac commented on src/main/java/attendance/controller/AttendanceMachine.java:
```diff
@@ -0,0 +1,136 @@
+        inputView.closeScanner();
+    }
+
+    //TODO : 객체지향 생활 원칙에서 else를 사용하지 말라고 하였는데, else if도 사용하지 말아야 하는 것인지 궁금합니다.
```
넵. else나 else if나 사용성을 생각하면 완벽하게 동일합니다.

bowook replied:
혹시, else if를 사용하지 않는 이유가 직관적이지 못해서 인가요?
else if를 해석하게 되면 
`첫 if 조건이 아니면서 else if 조건이 맞다면 실행` 이러한 방식으로 해석이 되어서 그런건지 궁금합니다!

ddaaac replied:
맞습니다. 조건이 연쇄적으로 중첩되다보면 읽는 호흡이 길어지게 되겠죠.

다만, 적절하게 쓸 데는 쓰는 게 좋다고 생각합니다. 우테코 레벨 1 미션에서는 요구사항이 else를 쓰지 않는 것이니 반드시 지켜줘야하는 거구요.

ddaaac commented on src/main/java/attendance/controller/AttendanceMachine.java:
```diff
@@ -0,0 +1,136 @@
+    private void confirmAttendance(LocalDate now, Crews crews, Register register) {
+        Crew crew = findCrew(crews);
+        List<String> attendanceTime = List.of(findAttendanceTime().split(":"));
+        LocalDateTime localDateTime = LocalDateTime.of(now.getYear(), now.getMonthValue(), now.getDayOfMonth(), Integer.parseInt(attendanceTime.get(0)), Integer.parseInt(attendanceTime.get(1)));
```
`LocalDate#atTime()`을 활용해볼까요?

bowook replied:
출석 수정에 대해서는 날짜를 입력받아야 하니까 적용 못하고, 출석 확인에 대해서는 시간만 전달 받는 것이니까
now.atTime()을 하면 되게 좋네요!
처음 알게된 기능입니다 감사합니다!

ddaaac commented on src/main/java/attendance/domain/DateInfo.java:
```diff
@@ -0,0 +1,103 @@
+public class DateInfo {
+
+    private LocalDateTime localDateTime;
+    private AttendanceStatus attendanceStatus;
```
여기도 DateInfo와 같은 추상적인 네이밍보다는 출결이라는 개념이 포함되게 이름을 지어주는 건 어떨까요?

bowook replied:
AttendanceChecker로 수정하였습니다!

ddaaac commented on src/main/java/attendance/domain/DateInfo.java:
```diff
@@ -0,0 +1,103 @@
+        return 0;
+    }
+
+    private AttendanceStatus calculateStatus() {
```
출결 상태를 결정하는 로직은 `AttendanceStatus`에 두는게 적절하지 않을까요?

bowook replied:
동의합니다!
출결을 확인하는 곳에서 출결 상태를 결정할 책임이 없는 것 같아요!
AttendanceStatus에게 출석 시간을 메시지로 전달을 하고, AttendanceStatus에서 메시지를 받은 후 해당 되는 출결 상태를 반환해 주는 방식으로 하면 객체의 책임이 명확해지는 것 같아요!

ddaaac commented on src/main/java/attendance/view/OutputView.java:
```diff
@@ -0,0 +1,161 @@
+                calculateTotalAbsence(e1.getValue())
+        ));
+        Set<Crew> excludedCrews = orderByAbsence(crewList);
+        orderByNickname(crewList, excludedCrews);
```
몇 가지 얘기를 해볼 수 있겠네요.

- 요구사항에 적혀있는 내용들은 도메인 로직이라고 보면 됩니다. View에 도메인 로직을 담기 적당할까요?
- 또, 정렬이 잘 되는지는 어떻게 확인해야할까요? 참고로 View는 테스트를 작성할 필요가 없습니다.

bowook replied:
View에 도메인 로직을 담는 것은 적절하지 않다고 생각합니다.


다만, 현재 출석부 미션에서는 도메인 로직에서 정렬된 상태를 가지고 어떠한 역할을 수행하는 것이 아니였고, 단순히 사용자에게 정렬된 상태로 보여주기 위함이라고 이해했습니다.


그러면 도메인 로직에서 정렬을 하는 것 보다 도메인 로직은 제적 위험자 크루들을 전달을 해주고 뷰에서 정렬을 해 보여주는 것이 적절하다고 생각했어요


만일, 도메인 상에서 객체가 정렬된 상태를 가지고 `제적 위험자 1순위 대상을 찾는다`와 비슷한 역할을 수행해야 한다면 도메인 상에서 정렬을 하는 것이 적절하다고 생각해요

ddaaac replied:
그러면 정렬 상태에 대한 테스트를 해볼 필요가 없다는 얘기인걸까요?
아니면 View에 테스트를 작성하신다는 걸까요?

bowook replied:
다시 곰곰히 생각해봤는데,
정렬을 테스트를 해야한다고 생각을 하였을 때, 도메인에서 정렬된 상태를 반환해주고, 뷰에서는 반환된 값을 받아서 출력 형식에 맞게  출력을 해주면,
정렬이 올바르게 되었는지 테스트도 할 수 있을 것 같아서 적절하다는 생각이 드네요.

ddaaac commented on src/main/java/attendance/view/OutputView.java:
```diff
@@ -0,0 +1,161 @@
+    private int calculateTotalAbsence(List<Integer> absenceCounts) {
+        int absence = absenceCounts.get(0);
+        int late = absenceCounts.get(1);
+        return absence + (late / 3);
```
여기도 마찬가지입니다. "지각 3번을 결석으로 간주한다"는 도메인 요구사항입니다.

bowook replied:
지각을 결석으로 간주해서 정렬이라, 나누기 3 없이 absence와 late를 더해서 졍렬하도록 수정하였습니다.

제적 대상자를 `판단`할 때는 지각 3번을 결석으로 간주해서 제적 대상자를 뽑아내는 것이 도메인 로직이라고 생각했어요!

ddaaac reviewed (changes requested):
안녕하세요 우가 👋 

리뷰 반영 잘 해주셨네요.

몇 가지 코멘트 남겼으니 확인해주세요.

bowook committed:
refactor(Crews) : 인덱스를 통해 접근하는 것이 아닌 getFirst() 사용

bowook committed:
refactor(Crews, Register) : 객체 생성 분할

AttendanceParser라는 유틸성 클래스를 만들어서 라인을 읽을 때 바로 Crews와 dateTime을 가공할 수 있도록 하였습니다. 이후, 객체 생성 부분은 AttendanceParser의 책임으로 하였습니다.

bowook committed:
refactor(AttendanceMachine) : now.atTime() 으로 수정

bowook committed:
refactor(DateInfo) : AttendacneChecker로 클래스명 수정

bowook committed:
refactor(AttendanceChecker) : boolean을 반환하도록 수정

객체지향적인 관점에서 바라볼 수 있도록 연습을 많이 해야겠다고 느끼게 도었습니다. 출결을 확인하는 객체는 출결을 했다, 안 했다만 반환해 주면 되는 책임이 있고, 이를 호출하는 클라이언트에서 판단해서 1과 0을 더해주는 것으로 수정하였습니다.

bowook committed:
refactor(AttendanceStatus) : 출결 상태를 결정하는 로직을 수정하였습니다.

bowook committed:
refactor(OutputView) : 지각을 결석으로 간주해서 정렬

bowook committed:
test(RegisterTest) : 테스트 추가

bowook committed:
test(AttendanceStatus) : 주말 반환 확인

bowook committed:
refactor(Register) : Crew 존재 여부 확인

ddaaac commented on src/main/java/attendance/controller/AttendanceMachine.java:
```diff
@@ -0,0 +1,137 @@
+        String beforeStatus = beforeInfo.getAttendanceStatus();
+        AttendanceChecker modifiedInfo = register.modifyInfo(crew, localDateTime);
+        outputView.writeAttendanceModifyCheck(beforeHour, beforeMinute, beforeStatus, modifiedInfo);
+    }
```
이 메서드가 하는 일이 너무 많은 것 같아요. 한 눈에 들어오지도 않구요.
적절히 메서드를 나누거나, 다른 객체에 책임을 넘기는 것도 좋아보이네요.

bowook replied:
초기에 변경할 Target 시간을 LocalDateTime으로 만들어주는 메서드로 분리하였습니다!

ddaaac commented on src/main/java/attendance/controller/AttendanceMachine.java:
```diff
@@ -0,0 +1,137 @@
+    private void confirmAttendance(LocalDate now, Crews crews, Register register) {
+        Crew crew = findCrew(crews);
+        List<String> attendanceTime = List.of(findAttendanceTime().split(":"));
+        LocalDateTime localDateTime = now.atTime(Integer.parseInt(attendanceTime.getFirst()), Integer.parseInt(attendanceTime.getLast()));
```
여기도 위와 중복된 내용들이 존재하는 것 같습니다.
적절하게 처리해보면 좋을 것 같아요.

ddaaac commented on src/main/java/attendance/domain/AttendanceChecker.java:
```diff
@@ -0,0 +1,52 @@
+        this.attendanceStatus = AttendanceStatus.calculateStatus(localDateTime);
+    }
+
+    public static AttendanceChecker of(LocalDateTime localDateTime) {
```
이 둘은 어떤 차이가 있나요?

bowook replied:
기본 생성자를 통해 생성하는 것 보다 정적 팩토리 메서드를 사용해서 생성하면 생성할 때 단순 new가 아닌 of나 from과 같은의미를 넣을 수 있어서 사용했습니다.

ddaaac replied:
new와 of에 어떤 차이가 있는지 잘 와닿지가 않네요. 사용하는 입장에서 크게 차이를 느낄까요?

bowook replied:
단순한 객체라서 new를 사용한다고 해도 기능적으로 차이가 없어서 사용자 입장에서는 차이를 느끼기 어려울 것 같네요

ddaaac commented on src/main/java/attendance/controller/AttendanceMachine.java:
```diff
@@ -0,0 +1,137 @@
+        Crew crew = findModifiyCrew(crews);
+        int modifyDate = readModifyDay();
+        List<String> modifyTime = List.of(findModifyTime().split(":"));
+        LocalDateTime localDateTime = LocalDateTime.of(now.getYear(), now.getMonthValue(), modifyDate, Integer.parseInt(modifyTime.getFirst()), Integer.parseInt(modifyTime.getLast()));
```
`now.withDayOfMonth(-).atTime(-)` 이런 식으로 체이닝해서 쓸 수 있습니다.

bowook replied:
감사합니다

ddaaac commented on src/main/java/attendance/domain/AttendanceRegistry.java:
```diff
@@ -0,0 +1,98 @@
+    private static List<AttendanceChecker> makeDefaultDateInfos(LocalDate now) {
+        List<AttendanceChecker> attendanceCheckers = new ArrayList<>();
+        for (int day = 1; day <= now.getDayOfMonth(); day++) {
+            LocalDate currentDay = LocalDate.of(now.getYear(), now.getMonthValue(), day);
```
`now.withDayOfMonth()` 여기도 쓸 수 있겠네요.

bowook replied:
감사합니다.

ddaaac commented on src/main/java/attendance/domain/AttendanceRegistry.java:
```diff
@@ -0,0 +1,98 @@
+    private static void addWeekdayDateInfo(LocalDate currentDay, List<AttendanceChecker> attendanceCheckers) {
+        if (checkHoliday(currentDay)) return;
+
+        attendanceCheckers.add(AttendanceChecker.makeDefaultValue(currentDay.getYear(),currentDay.getMonthValue(),
+                currentDay.getDayOfMonth()));
```
파라미터를 3개 넘길 필요없이 LocalDate만 넘겨도 되지 않을까요?

bowook replied:
감사합니다.

ddaaac commented on src/main/java/attendance/domain/AttendanceRegistry.java:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    private static boolean checkHoliday(LocalDate currentDay) {
+        return currentDay.getDayOfWeek().getValue() >= 6;
```
숫자 비교는 직관적인 것 같지는 않네요. 토요일 또는 일요일인지 직접 비교하게 구현해도 좋을 것 같아요.

bowook replied:
Weekday를 이용해서 할 수 있을 것 같아요!

ddaaac commented on src/main/java/attendance/domain/AttendanceRegistry.java:
```diff
@@ -0,0 +1,98 @@
+        return currentDay.getDayOfWeek().getValue() >= 6;
+    }
+
+    public AttendanceChecker findByDate(int date) {
```
date와 day를 명확히 구분하면 좋을 것 같네요. 용어들이 혼용돼서 사용되기 시작하면 코드 읽기가 복잡해집니다.

int와 같은 원시타입보다는 LocalDate와 같은 포장된 값을 사용하는 것도 하나의 방법이구요.

bowook replied:
감사합니다.

ddaaac commented on src/main/java/attendance/domain/Register.java:
```diff
@@ -0,0 +1,75 @@
+
+    public Register(Crews crews, LocalDate now) {
+        register = new HashMap<>();
+        crews.register(register, now);
```
본인이 가진 컬렉션을 다른 객체에 넘기고, 그 객체에서 컬렉션을 조작하게 된다면

이 객체 입장에서는 다른 객체에서 이 컬렉션을 조작했는지 하지 않았는지 알 방법이 있을까요?

일급 컬렉션을 왜 사용하는지 생각해보시면 좋을 것 같습니다.

bowook replied:
말씀해주셔서 감사합니다!
고민을 해봤는데, 말씀해 주신 부분이 crews 컬렉션에 대한 불변성 보장을 못하니까 발생하는 문제라고 이해를 했어요.
기존 코드처럼 Crews가 불변성이 보장되지 않는다고 한다면, 다른 객체에서 조작을 하였을  때, Crews 를 이용하고 있는 Register는 알 수 있는 방법이 없다고 생각합니다.
이러한 부분을 해결하기 위해서 불변성을 보장할 수 있도록 수정하였습니다.
감사합니다

ddaaac replied:
이 부분 반영된걸까요? Map register를 여전히 crew에서 변경하고 있는 것 같네요.

bowook replied:
말씀해주신 피드백을 잘못 이해한 것 같아요, Crews를 조작할 수 없도록 하라는 피드백인 줄 알았습니다.
Crews가 Register의 내부 구조를 직접 변경하는 권한을 가지게 되는 문제가 있었네요!
감사합니다.

ddaaac reviewed (changes requested):
안녕하세요 우가 👋 

몇 가지 코멘트 남겼으니 확인해주세요.

bowook committed:
refactor(AttendanceMachine) : 메서드 책임 분리

bowook committed:
refactor(AttendanceMachine) : 출석 확인 기능 메서드 분리

bowook committed:
refactor(AttendanceChecker, AttendanceRegistry) : LocalDateTime을 넘기도록 수정

bowook committed:
refactor(AttendanceRegistry) : checkHoliday 직관적이게 수정

bowook committed:
refactor(AttendanceRegistry) :  date -> day로 명칭 수정

bowook committed:
refactor(Crews) : 불변성 보장

bowook committed:
refactor(AttendanceRegistry) : findByDay 원시값이 아닌 LocalDateTime 사용

bowook committed:
refactor(Register) : 정렬을 도메인에서 수행하도록 수정

bowook committed:
refactor(Register) : 인덱스를 통해 get이 아닌, getFirst와 getLast 사용

bowook committed:
test(Register) : 정렬 테스트 추가

bowook committed:
feat(Register) : 캠퍼스 운영 시간 제한 설정

bowook committed:
test(RegisterTest) : 캠퍼스 운영 시간 검증 테스트

bowook committed:
refactor(AttendanceRegistry) : 출석 기록 확인 기능 Domain으로 분리

bowook committed:
refactor(Register) : CrewStatus를 사용하 수 있도록 조건문 변경

bowook committed:
refactor(AttendanceRegistry) : 출석 기록 계산 수정

ddaaac commented on src/main/java/attendance/controller/AttendanceMachine.java:
```diff
@@ -0,0 +1,148 @@
+
+    private String findModifyTime() {
+        return retryUntilValidInput(inputView::readModifyTime);
+    }
```
time 같은건 여기서 미리 변환해두면 사용하기 편하지 않을가 싶네요.

ddaaac commented on src/main/java/attendance/domain/Register.java:
```diff
@@ -0,0 +1,107 @@
+        }
+    }
+
+    public List<Entry<Crew, List<Integer>>> findAllExpertRiskCrews() {
```
메서드의 시그니처는 명확해야합니다. 시그니처만 보고도 의미를 파악하기 쉬워야하죠.
- `List<Integer>`가 어떤 값인지 파악이 어렵습니다.
- `List<Entry<>>`를 더 쉽게 표현할 수 있는 자료구조가 있을 것 같아요. 자료구조의 중첩은 가독성을 떨어트립니다.

bowook replied:
List<>를 이용하도록 수정하였습니다.!
기존에, 닉네임으로 정렬이 되고 있지 않는 문제도 있어서, 결석 횟수로 정렬 하였을 때의 값이 0을 반환할 경우에는 닉네임 순으로 정렬할 수 있도록 수정하였습니다!
List<CrewRisk>로 관리하였습니다.

ddaaac reviewed (changes requested):
안녕하세요 우가 👋 

몇 가지 코멘트 남겨두었으니 확인해주세요.

bowook committed:
refactor(AttendanceMachine) : 시간 변환 수정

bowook committed:
refactor(AttendanceChecker) : 정적 팩토리 메서드에서 그냥 생성자로 객체 생성하도록 수정

bowook committed:
refactor(Register) : Register 내부구조 변경 권한을 Register만 가지도록 수정

bowook committed:
refactor(Register) : 정렬 반환 수정

bowook committed:
refactor(Register) : 닉네임 순으로 정렬이 가능하도록 수정

bowook committed:
chore(Register) : 주석 삭제

ddaaac commented on src/main/java/attendance/controller/AttendanceMachine.java:
```diff
@@ -0,0 +1,148 @@
+        Crew crew = findModifiyCrew(crews);
+        LocalDateTime targetDateTime = readModifyDateTime(now);
+        AttendanceChecker beforeInfo = register.findInfo(crew, targetDateTime);
+        LocalTime beforeTime = LocalTime.of(beforeInfo.getLocalDateTime().getHour(), beforeInfo.getLocalDateTime().getMinute());
```
여기도 beforeInfo 에서 값을 get 하는 대신 beforeInfo에 메세지를 던질 수 있을 것 같네요.

bowook replied:
감사합니다,
메시지를 전달하는 방식으로 적용해보겠습니다.

ddaaac commented on src/main/java/attendance/controller/AttendanceMachine.java:
```diff
@@ -0,0 +1,148 @@
+        return retryUntilValidInput(() -> crews.findCrew(inputView.readModifyCrewName()));
+    }
+
+    private List<String> findAttendanceTime() {
+        return Parser.convertToGroup(retryUntilValidInput(inputView::readAttendanceTime));
+    }
+
+    private List<String> findModifyTime() {
+        return Parser.convertToGroup(retryUntilValidInput(inputView::readModifyTime));
+    }
```
https://github.com/woowacourse/java-attendance/pull/73#discussion_r1969893354 코멘트와 동일하게 여기도 Time으로 만들어서 반환하면 좋을 것 같네요.

코멘트를 보고 그 곳에만 적용하는 대신 다른 곳에도 적용할 곳이 있나 살펴보면 좋을 것 같아요.

bowook replied:
위 코멘트와 동일한 코멘트인 곳인데, 제가 피드백을 잘못 이해한 것 같습니다.
적용해보겠습니다!

ddaaac commented on src/main/java/attendance/domain/AttendanceRegistry.java:
```diff
@@ -0,0 +1,94 @@
+        return new AttendanceRegistry(attendanceCheckers);
+    }
+
+    public void calculateAttendanceHistory() {
+        List<Integer> attendanceCounts = new ArrayList<>(List.of(0, 0, 0));
+        for (AttendanceChecker attendanceChecker : attendanceCheckers) {
+            attendanceCounts.set(0, attendanceCounts.getFirst() + mappingStatusToNumber(attendanceChecker.isAbsence()));
+            attendanceCounts.set(1, attendanceCounts.get(1) + mappingStatusToNumber(attendanceChecker.isLate()));
+            attendanceCounts.set(2, attendanceCounts.getLast() + mappingStatusToNumber(attendanceChecker.isAttendance()));
+        }
+        this.attendanceTraces = attendanceCounts;
+        this.crewStatus = CrewStatus.from(attendanceCounts.getLast(), attendanceCounts.getFirst());
```
attendanceTraces와 crewStatus를 메서드에서 반환하는 것과, 필드 세팅하는 방식은 어떤 차이가 있을까요?

bowook replied:
필드로 세팅하는 방식은 여러 행위를 통해서 상태가 변할 때 필드로 세팅하고,

그게 아닌, 메서드에서 바로 반환하는 것은 한 행위만으로 관리되는 값이면 필드로 두지 않고 반환한다는 차이가 있다고 생각했습니다.

ddaaac reviewed (approved):
안녕하세요 우가 👋 

다음 단계 진행해주세요
